From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Denis Grigorev <d.grigorev@omprussia.ru>
Date: Wed, 17 Jun 2026 12:55:21 +0200
Subject: Add a video decoder based on gecko-camera. JB#56755

Adds the Gecko integration for a video decoder plugin that makes use of gecko-camera. The decoder implementation lives in the package-level gecko-camera-decoder tree and is linked into dom/media/platforms for Gecko builds.
---
 dom/media/platforms/PDMFactory.cpp       | 13 +++++++++++++
 dom/media/platforms/gecko-camera         |  1 +
 dom/media/platforms/moz.build            | 11 +++++++++++
 modules/libpref/init/StaticPrefList.yaml | 13 +++++++++++++
 4 files changed, 38 insertions(+)
 create mode 120000 dom/media/platforms/gecko-camera

diff --git a/dom/media/platforms/PDMFactory.cpp b/dom/media/platforms/PDMFactory.cpp
index 49d9e835f74a..b1c2186e2ad9 100644
--- a/dom/media/platforms/PDMFactory.cpp
+++ b/dom/media/platforms/PDMFactory.cpp
@@ -64,6 +64,10 @@
 #ifdef MOZ_OMX
 #  include "OmxDecoderModule.h"
 #endif
+#ifdef MOZ_ENABLE_WEBRTC_GECKOCAMERA
+#  include "gecko-camera/GeckoCameraDecoderModule.h"
+#endif
+
 
 #include <functional>
 
@@ -194,6 +198,9 @@ class PDMInitializer final {
 #endif
 #ifdef MOZ_FFMPEG
     FFmpegRuntimeLinker::Init();
+#endif
+#ifdef MOZ_ENABLE_WEBRTC_GECKOCAMERA
+    GeckoCameraDecoderModule::Init();
 #endif
   }
 
@@ -735,6 +742,12 @@ void PDMFactory::CreateDefaultPDMs() {
                StaticPrefs::media_android_media_codec_preferred());
   }
 #endif
+#ifdef MOZ_ENABLE_WEBRTC_GECKOCAMERA
+  if (StaticPrefs::media_gecko_camera_codec_enabled()) {
+    StartupPDM(GeckoCameraDecoderModule::Create(),
+               StaticPrefs::media_gecko_camera_codec_preferred());
+  }
+#endif
 
   CreateAndStartupPDM<AgnosticDecoderModule>();
 
diff --git a/dom/media/platforms/gecko-camera b/dom/media/platforms/gecko-camera
new file mode 120000
index 000000000000..215967f59e8c
--- /dev/null
+++ b/dom/media/platforms/gecko-camera
@@ -0,0 +1 @@
+../../../../gecko-camera-decoder
\ No newline at end of file
diff --git a/dom/media/platforms/moz.build b/dom/media/platforms/moz.build
index c71f22a22eee..776b83648287 100644
--- a/dom/media/platforms/moz.build
+++ b/dom/media/platforms/moz.build
@@ -135,6 +135,17 @@ if CONFIG["MOZ_WIDGET_TOOLKIT"] == "android":
         "/media/libyuv/libyuv/include",
     ]
 
+if CONFIG["MOZ_ENABLE_WEBRTC_GECKOCAMERA"]:
+    EXPORTS += [
+        "gecko-camera/GeckoCameraDecoderModule.h",
+    ]
+    UNIFIED_SOURCES += [
+        "gecko-camera/GeckoCameraDecoderModule.cpp",
+        "gecko-camera/GeckoCameraVideoDecoder.cpp",
+    ]
+    OS_LIBS += CONFIG["LIBGECKOCAMERA_LIBS"]
+    CXXFLAGS += CONFIG["LIBGECKOCAMERA_CFLAGS"]
+
 FINAL_LIBRARY = "xul"
 
 # Add libFuzzer configuration directives
diff --git a/modules/libpref/init/StaticPrefList.yaml b/modules/libpref/init/StaticPrefList.yaml
index 13e75da963ac..c900d851198a 100644
--- a/modules/libpref/init/StaticPrefList.yaml
+++ b/modules/libpref/init/StaticPrefList.yaml
@@ -10206,6 +10206,19 @@
     mirror: always
 #endif  # ANDROID
 
+#ifdef MOZ_EMBEDLITE
+  # Enable the gecko-camera based decoder by default.
+-   name: media.gecko-camera-codec.enabled
+    type: RelaxedAtomicBool
+    value: true
+    mirror: always
+
+-   name: media.gecko-camera-codec.preferred
+    type: RelaxedAtomicBool
+    value: true
+    mirror: always
+#endif  # MOZ_EMBEDLITE
+
 #ifdef MOZ_OMX
 -   name: media.omx.enabled
     type: bool
